### PR TITLE
Support BFME water areas

### PR DESCRIPTION
**Pull request by feliwir**
Added support for water & wave areas. Support water bump textures

### CODE REVIEW

File: src/OpenSage.Game/Assets/Shaders/Water.frag [hunk 1]
```diff
@@ -27,7 +28,8 @@ layout(location = 0) out vec4 out_Color;
 
 void main()
 {
-    vec3 worldNormal = vec3(0, 0, 1);
+    vec2 waterUV = vec2(in_WorldPosition.x / 320, in_WorldPosition.y / 320);
+    vec3 worldNormal = texture(sampler2D(BumpTexture, Sampler), waterUV).xyz;
```
Comment (feliwir) [inline src/OpenSage.Game/Assets/Shaders/Water.frag]: This is maybe not mathematically correct

File: src/OpenSage.Game/Content/ScopedAssetCollection.cs [hunk 2]
```diff
@@ -73,6 +73,12 @@ public TAsset GetByName(string name)
             if (_loader != null)
             {
                 var newValue = _loader.Load(name, _assetStore.LoadContext);
+                //TODO: should this happen?
+                if(newValue == null)
```
Comment (feliwir) [inline src/OpenSage.Game/Content/ScopedAssetCollection.cs]: Without this we crash on most BFME maps

File: src/OpenSage.Game/Diagnostics/MainView.cs [hunk 3]
```diff
@@ -102,6 +102,10 @@ public void Draw(ref bool isGameViewFocused)
                     {
                         foreach (var mapCache in _context.Game.AssetStore.MapCaches)
                         {
+                            //TODO: we should probably cache the validity of entries
+                            if (_context.Game.ContentManager.FileSystem.GetFile(mapCache.Name) == null)
```
Comment (feliwir) [inline src/OpenSage.Game/Diagnostics/MainView.cs]: The mapcache contains invalid entries

File: src/OpenSage.Game/Graphics/Shaders/WaterShaderResources.cs [hunk 4]
```diff
@@ -46,12 +47,12 @@ internal sealed class WaterShaderResources : ShaderResourcesBase
                     RenderPipeline.GameOutputDescription)));
         }
 
-        public ResourceSet CreateMaterialResourceSet(Texture texture)
+        public ResourceSet CreateMaterialResourceSet(Texture waterTexture, Texture bumpTexture = null)
```
Comment (tgjones) [inline src/OpenSage.Game/Graphics/Shaders/WaterShaderResources.cs]: I don’t think you need “= null” here?

File: src/OpenSage.Game/Terrain/WaterArea.cs [hunk 5]
```diff
@@ -44,16 +44,43 @@ public sealed class WaterArea : DisposableBase
...
+                return false;
+            }
+
+            result = new WaterArea(loadContext, area);
+            return true;
+        }
 
+        internal static bool TryCreate(
+            AssetLoadContext loadContext,
+            StandingWaveArea area,
```
Comment (tgjones) [inline src/OpenSage.Game/Terrain/WaterArea.cs]: We may end up wanting a different class for waves vs static water, but this is fine for now.

File: src/OpenSage.Game/Graphics/Shaders/WaterShaderResources.cs (hunk 4 above)
Comment (feliwir) [reply src/OpenSage.Game/Graphics/Shaders/WaterShaderResources.cs]: Yep, fixed

**Pull request merged**